### PR TITLE
Separate subjects into categories

### PR DESCRIPTION
**Pull request by santiagorodriguez96**
## Why

A lot of subjects were showing up under optional subjects despite not really being optional subjects.

This PR tries to separate some of them into different categories such as `Reválidas`, `Materias Inactivas`, `Módulos de Taller y Extensión` or `Materias dictadas en institutos del interior`.

## How To Review

Commit by commit suggested

### CODE REVIEW

File: db/migrate/20230303213409_add_revalid_to_subjects.rb [hunk 1]
```diff
@@ -0,0 +1,5 @@
+class AddRevalidToSubjects < ActiveRecord::Migration[7.0]
+  def change
+    add_column :subjects, :revalid, :boolean, default: false
```
Comment (santiagorodriguez96) [inline db/migrate/20230303213409_add_revalid_to_subjects.rb:3]: We should probably reconsider the naming here...

I couldn't came up with anything so I went with this one instead 🙃 hoping that we would come up with something better later...

File: db/migrate/20230306135310_add_external_to_subjects.rb [hunk 2]
```diff
@@ -0,0 +1,5 @@
+class AddExternalToSubjects < ActiveRecord::Migration[7.0]
+  def change
+    add_column :subjects, :external, :boolean, default: false
```
Comment (santiagorodriguez96) [inline db/migrate/20230306135310_add_external_to_subjects.rb:3]: Same here lol (but here is even worse 😅 )

File: db/migrate/20230307130502_add_extension_module_to_subjects.rb [hunk 3]
```diff
@@ -0,0 +1,5 @@
+class AddExtensionModuleToSubjects < ActiveRecord::Migration[7.0]
+  def change
+    add_column :subjects, :extension_module, :boolean
```
Comment (santiagorodriguez96) [inline db/migrate/20230307130502_add_extension_module_to_subjects.rb:3]: Actually, subjects under this category are not only `Módulos de Extensión` but also `Módulos de Taller` so naming here is not reeeeeally accurate...

Let me know what you guys think about that 🙂

File: app/models/subject.rb [hunk 4]
```diff
@@ -22,5 +22,35 @@ def approved?(approved_approvable_ids)
...
+    elsif revalid?
+      :revalid
+    elsif external?
+      :external
+    elsif extension_module?
+      :extension_module
+    else
+      :optional
+    end
+  end
```
Comment (santiagorodriguez96) [inline app/models/subject.rb:49]: We could also avoid creating one attribute per each category and just add an enum attribute `category`.

Thoughts about using that approach instead?

File: app/helpers/subjects_helper.rb [hunk 5]
```diff
@@ -1,24 +1,41 @@
...
-      "Quinto"
-    when 6
-      "Sexto"
-    when 7
-      "Séptimo"
-    when 8
-      "Octavo"
-    when 9
-      "Noveno"
+  CATEOGIRES_SORT_ORDER = {
```
[the remark this answers is not in the document]
Comment (santiagorodriguez96) [reply app/helpers/subjects_helper.rb]: Thanks! Updated in [4c3101f](https://github.com/cedarcode/mi_carrera/pull/247/commits/4c3101fe5bac9421ab06374442aff0ec458b3e15) 🙂

File: app/helpers/subjects_helper.rb [hunk 6]
```diff
@@ -1,24 +1,41 @@
...
+
+  def subjects_grouped_by_category(subjects)
+    subjects.group_by(&:category).sort_by { |key, _value| CATEOGIRES_SORT_ORDER[key] }.to_h
+  end
+
+  def formatted_category(category)
+    case category
+    when :first_semester then 'Primer semestre'
+    when :second_semester then 'Segundo semestre'
+    when :third_semester then 'Tercer semestre'
```
[the remark this answers is not in the document]
Comment (santiagorodriguez96) [reply app/helpers/subjects_helper.rb]: I went for a different approach in the end, so this is not needed anymore 🙂 

Check it out: [4c3101f](https://github.com/cedarcode/mi_carrera/pull/247/commits/4c3101fe5bac9421ab06374442aff0ec458b3e15)

File: app/helpers/subjects_helper.rb [hunk 7]
```diff
@@ -1,24 +1,20 @@
...
+    when 'fourth_semester' then 'Cuarto semestre'
+    when 'fifth_semester' then 'Quinto semestre'
+    when 'sixth_semester' then 'Sexto semestre'
+    when 'seventh_semester' then 'Séptimo semestre'
+    when 'eighth_semester' then 'Octavo semestre'
+    when 'nineth_semester' then 'Noveno semestre'
+    when 'inactive' then 'Materias inactivas'
+    when 'optional' then 'Materias opcionales'
+    when 'revalid' then 'Reválidas'
+    when 'external' then 'Materias dictadas en institutos del interior'
```
Comment (RenzoMinelli) [inline app/helpers/subjects_helper.rb:16]: I would maybe change `external` to something like `outside_montevideo` or  something more descriptive.

File: app/helpers/subjects_helper.rb [hunk 8]
```diff
@@ -1,24 +1,20 @@
...
+    when 'extension_module' then 'Módulos de Taller y Extensión'
```
Comment (RenzoMinelli) [inline app/helpers/subjects_helper.rb:17]: I would leave `Taller` and `Extensión` without capital letters

File: app/helpers/subjects_helper.rb (hunk 7 above)
Comment (santiagorodriguez96) [reply app/helpers/subjects_helper.rb:16]: Sounds good to me! 👀

Comment (santiagorodriguez96) [reply app/helpers/subjects_helper.rb]: Fixed in [a3b64de](https://github.com/cedarcode/mi_carrera/pull/247/commits/a3b64de743f916b53c2781b376a6dd2958c199a2) 🙂

File: app/helpers/subjects_helper.rb (hunk 8 above)
Comment (santiagorodriguez96) [reply app/helpers/subjects_helper.rb]: Fixed in [829a10a](https://github.com/cedarcode/mi_carrera/pull/247/commits/829a10af135d1a6a77dbd624fb2f0d98c787999b) 🙂

**Pull request merged**